packages/core/src/render3/util/discovery_utils.ts

```diff
@@ -6,22 +6,52 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {ChangeDetectionStrategy} from '../../change_detection/constants';
 import {Injector} from '../../di/injector';
+import {ViewEncapsulation} from '../../metadata/view';
 import {assertEqual} from '../../util/assert';
 import {assertLView} from '../assert';
-import {discoverLocalRefs, getComponentAtNodeIndex, getDirectivesAtNodeIndex, getLContext} from '../context_discovery';
+import {discoverLocalRefs, getComponentAtNodeIndex, getDirectivesAtNodeIndex, getLContext, readPatchedLView} from '../context_discovery';
+import {getComponentDef, getDirectiveDef} from '../definition';
 import {NodeInjector} from '../di';
 import {buildDebugNode} from '../instructions/lview_debug';
 import {LContext} from '../interfaces/context';
 import {DirectiveDef} from '../interfaces/definition';
 import {TElementNode, TNode, TNodeProviderIndexes} from '../interfaces/node';
-import {isLView} from '../interfaces/type_checks';
-import {CLEANUP, CONTEXT, DebugNode, FLAGS, LView, LViewFlags, T_HOST, TVIEW, TViewType} from '../interfaces/view';
+import {isComponentHost, isLView, isTNode} from '../interfaces/type_checks';
+import {CLEANUP, CONTEXT, DebugNode, FLAGS, HOST, LView, LViewFlags, T_HOST, TVIEW, TViewType} from '../interfaces/view';
+
 import {stringifyForError} from './stringify_utils';
 import {getLViewParent, getRootContext} from './view_traversal_utils';
 import {getTNode, unwrapRNode} from './view_utils';
 
 
+/**
+ * Checks if a specified DOM element is a host element.
+ *
+ * @param element DOM element to verify if it's a host element.
+ * @returns If the element is a host element or not.
+ */
+function isComponentHostElement(element: Element): boolean {
+  const lView = readPatchedLView(element);
+
+  // All host elements have LViews associated with them. If we can't find patched
+  // data we can safely assume the current element is not a host.
+  if (!isLView(lView)) {
+    return false;
+  }
+  const tView = lView[TVIEW];
+  const tData = tView.data;
+  const components = tView.components;
+  return components?.some(component => {
+    const tNode = tData[component];
+    if (!tNode || !isTNode(tNode) || lView[component][HOST] !== element) {
```

Comment:
`isTNode()` already checks whether `tNode` is truthy.

```suggestion
    if (!isTNode(tNode) || lView[component][HOST] !== element) {
```